### PR TITLE
Let WalkingEntities be pushed by water

### PR DESCRIPTION
**Pull request by FlamingKnight**
It isn't probably the best though works nicely for now, hope that this helps!

### CODE REVIEW

File: src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java [hunk 1]
```diff
@@ -46,8 +48,8 @@ protected void checkTarget() {
 
         double near = Integer.MAX_VALUE;
 
-        for (Entity entity : this.getLevel().getEntities()) {
-            if (entity == this || !(entity instanceof EntityCreature) || !this.canTarget(entity)) {
+        for (Entity entity : this.getViewers().values()) {
+            if (entity == this || !(entity instanceof EntityCreature) || entity instanceof Animal) {
```
Comment (PetteriM1) [inline src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java:52]: Conflicts with https://github.com/Nukkit-coders/MobPlugin/commit/a89840a

File: src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java [hunk 2]
```diff
@@ -66,9 +68,10 @@ protected void checkTarget() {
             this.moveTime = 0;
             this.followTarget = creature;
             if (this.route == null && this.passengers.isEmpty()) this.target = creature;
+
```
Comment (PetteriM1) [inline src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java:71]: Why?

File: src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java [hunk 3]
```diff
@@ -66,9 +68,10 @@ protected void checkTarget() {
             this.moveTime = 0;
             this.followTarget = creature;
             if (this.route == null && this.passengers.isEmpty()) this.target = creature;
+
         }
 
-        if (this.followTarget instanceof EntityCreature && !this.followTarget.closed && this.followTarget.isAlive() && this.targetOption((EntityCreature) this.followTarget, this.distanceSquared(this.followTarget)) && this.target != null) {
+        if (this.followTarget instanceof EntityCreature && !((EntityCreature) this.followTarget).closed && this.followTarget.isAlive() && this.targetOption((EntityCreature) this.followTarget, this.distanceSquared(this.followTarget)) && this.target != null) {
```
Comment (PetteriM1) [inline src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java:74]: Why?

File: src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java [hunk 4]
```diff
@@ -97,8 +100,8 @@ protected void checkTarget() {
     protected boolean checkJump(double dx, double dz) {
         if (this.motionY == this.getGravity() * 2) {
             int b = level.getBlockIdAt(NukkitMath.floorDouble(this.x), (int) this.y, NukkitMath.floorDouble(this.z));
-            return b == BlockID.WATER || b == BlockID.STILL_WATER;
-        } else if (!(this instanceof SkeletonHorse)) {
```
Comment (PetteriM1) [inline src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java:101]: Was this intentional?

File: src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java [hunk 5]
```diff
@@ -154,20 +157,29 @@ public Vector3 updateMove(int tickDiff) {
                 return null;
             }
 
+            boolean inWater = getLevelBlock().getId() == 8 || getLevelBlock().getId() == 9;
+            int downFaceID = getLevelBlock().getSide(BlockFace.DOWN).getId();
```
Comment (PetteriM1) [inline src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java:161]: You should avoid multiple `getLevelBlock` calls if the block hasn't changed.

**Review by PetteriM1 — CHANGES_REQUESTED** [review]

File: src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java (hunk 4 above)
Comment (FlamingKnight) [reply src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java:101]: Nope, meant to uncomment it later

File: src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java [hunk 6]
```diff
@@ -154,20 +157,30 @@ public Vector3 updateMove(int tickDiff) {
...
 
                 double diff = Math.abs(x) + Math.abs(z);
-                if (this.stayTime > 0 || this.distance(this.followTarget) <= (this.getWidth()) / 2 + 0.05) {
+                if (!inWater && (this.stayTime > 0 || this.distance(this.followTarget) <= (this.getWidth()) / 2 + 0.05)) {
                     this.motionX = 0;
                     this.motionZ = 0;
                 } else {
-                    if (Utils.entityInsideWaterFast(this)) {
+                    if (getLevelBlock().getId() == 8) {
+                        BlockWater blockWater = (BlockWater) getLevelBlock();
```
Comment (PetteriM1) [inline src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java:176]: You can use getLevelBlock from line 160

File: src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java [hunk 7]
```diff
@@ -183,14 +196,19 @@ public Vector3 updateMove(int tickDiff) {
...
 
                 double diff = Math.abs(x) + Math.abs(z);
-                if (this.stayTime > 0 || this.distance(this.target) <= ((this.getWidth()) / 2 + 0.05) * nearbyDistanceMultiplier()) {
+                if (!inWater && (this.stayTime > 0 || this.distance(this.target) <= ((this.getWidth()) / 2 + 0.05) * nearbyDistanceMultiplier())) {
                     this.motionX = 0;
                     this.motionZ = 0;
                 } else {
-                    if (Utils.entityInsideWaterFast(this)) {
+                    if (getLevelBlock().getId() == 8) {
+                        BlockWater blockWater = (BlockWater) getLevelBlock();
```
Comment (PetteriM1) [inline src/main/java/nukkitcoders/mobplugin/entities/WalkingEntity.java:204]: You can use getLevelBlock from line 160

**Comment by PetteriM1** [comment]
Their speed in flowing water seems to be pretty high compared to vanilla. Maybe it should be decreased a little.

**Pull request merged**